### PR TITLE
OCPBUGS-17157: cache/json: use shared buffers for JSON decoding

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This commit uses a shared buffer in a shared decoder for the ListBundles call, which should reduce our memory footprint when we're asked for this data.

I also changed the sorting of the key-set to be an explicit call to sort.Slice, instead of an implicit side-effect from sets.New().List().

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Local testing only shows modest improvements, but these will be true regardless of GC performance, whereas `os.ReadFile()` can only perform this well when called N times if we have time to GC.

```
$ go tool pprof -diff_base /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.015.pb.gz /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz
File: opm
Build ID: f7be9c041bb04f86b16d09314008a9af727f3b81
Type: inuse_space
Time: Sep 6, 2023 at 9:26am (MDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top 
Showing nodes accounting for -571.39kB, 7.90% of 7234.88kB total
Showing top 10 nodes out of 98
      flat  flat%   sum%        cum   cum%
-1319.97kB 18.24% 18.24% -1319.97kB 18.24%  os.ReadFile
 1184.27kB 16.37%  1.88%  1184.27kB 16.37%  encoding/json.(*Decoder).refill
-1025.63kB 14.18% 16.05% -1025.63kB 14.18%  regexp/syntax.(*compiler).inst
 -627.57kB  8.67% 24.73%  -627.57kB  8.67%  encoding/json.unquoteBytes
  600.16kB  8.30% 16.43%   -27.41kB  0.38%  encoding/json.(*decodeState).literalStore
  557.26kB  7.70%  8.73%   557.26kB  7.70%  github.com/russross/blackfriday/v2.init
  553.04kB  7.64%  1.08%   553.04kB  7.64%  github.com/gogo/protobuf/proto.RegisterType
  536.37kB  7.41%  6.33%   536.37kB  7.41%  k8s.io/apimachinery/pkg/api/meta.init
 -516.01kB  7.13%   0.8%  -516.01kB  7.13%  regexp/syntax.appendRange
 -513.31kB  7.09%  7.90%  -513.31kB  7.09%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
```

File: pkg/cache/json.go [hunk 1]
```diff
@@ -58,31 +60,66 @@ func (q *JSON) ListBundles(ctx context.Context) ([]*api.Bundle, error) {
...
+			if err := file.Close(); err != nil {
+				logrus.WithError(err).WithField("file", file.Name()).Warn("could not close file")
 			}
 		}
+	}()
+	multiReader := io.MultiReader(readers...)
+	decoder := json.NewDecoder(multiReader)
+	index := 0
+	for {
+		var bundle api.Bundle
```
Comment (joelanford) [inline pkg/cache/json.go:105]: Is it possible/would it help/is it safe (given we pass a pointer to `Send`) to declare `bundle` this outside the loop and reuse it?

File: pkg/cache/json.go [hunk 2]
```diff
@@ -58,31 +60,66 @@ func (q *JSON) ListBundles(ctx context.Context) ([]*api.Bundle, error) {
...
+		return keys[i].name < keys[j].name
+	})
+	var files []*os.File
+	var readers []io.Reader
+	for _, key := range keys {
+		filename, ok := q.apiBundles[key]
+		if !ok {
+			return fmt.Errorf("package %q, channel %q, key %q not found", key.pkgName, key.chName, key.name)
+		}
+		file, err := os.Open(filename)
```
Comment (joelanford) [inline pkg/cache/json.go:87]: This means we'd have open files for every bundle in the catalog at once, so in the 1000s roughly. Is that reasonable? If that's undesirable, we could in theory have an `io.Pipe` with the pipe reader passed into the decoder. And then a separate goroutine opens and writes a file at a time into the pipe.

File: pkg/cache/json.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/cache/json.go:105]: If `api.Bundle` has an optional field and on `index=0` we set it, but `index=1` does not ... what does the content of the variable look like?

Comment (stevekuznetsov) [reply pkg/cache/json.go:105]: If it's logically correct, this is a good idea - we see an even larger reduction with it:

```
$ go tool pprof -diff_base /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.018.pb.gz
File: opm
Build ID: 6f0d15b7c8b7bcd22779237e2d78142137ab2ef4
Type: inuse_space
Time: Sep 6, 2023 at 9:26am (MDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for -662.18kB, 9.23% of 7175.57kB total
Dropped 2 nodes (cum <= 35.88kB)
Showing top 10 nodes out of 79
      flat  flat%   sum%        cum   cum%
-1114.54kB 15.53% 15.53% -1114.54kB 15.53%  encoding/json.(*decodeState).literalStore
 -557.26kB  7.77% 23.30%  -557.26kB  7.77%  github.com/russross/blackfriday/v2.init
 -536.37kB  7.48% 30.77%  -536.37kB  7.48%  k8s.io/apimachinery/pkg/api/meta.init
  518.02kB  7.22% 23.55%   518.02kB  7.22%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
     515kB  7.18% 16.38%      515kB  7.18%  regexp.makeOnePass.func1
     513kB  7.15%  9.23%      513kB  7.15%  golang.org/x/net/http2/hpack.newInternalNode
  512.38kB  7.14%  2.09%   512.38kB  7.14%  github.com/operator-framework/operator-registry/cmd/opm/index.newIndexDeprecateTruncateCmd
 -512.28kB  7.14%  9.23% -1626.82kB 22.67%  github.com/operator-framework/operator-registry/pkg/cache.(*JSON).Load
 -512.16kB  7.14% 16.36%  -512.16kB  7.14%  net.parseNSSConf
  512.02kB  7.14%  9.23%   512.02kB  7.14%  k8s.io/api/batch/v1.init
```

Comment (stevekuznetsov) [reply pkg/cache/json.go:105]: Seems to lead to a 500KiB savings but the source is unclear ... 

```
$ go tool pprof -diff_base /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.019.pb.gz
File: opm
Build ID: 6f0d15b7c8b7bcd22779237e2d78142137ab2ef4
Type: inuse_space
Time: Sep 6, 2023 at 9:26am (MDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for -662.16kB, 9.23% of 7175.57kB total
Dropped 5 nodes (cum <= 35.88kB)
Showing top 10 nodes out of 81
      flat  flat%   sum%        cum   cum%
-1114.54kB 15.53% 15.53%  -602.49kB  8.40%  encoding/json.(*decodeState).literalStore
 -557.26kB  7.77% 23.30%  -557.26kB  7.77%  github.com/russross/blackfriday/v2.init
 -536.37kB  7.48% 30.77%  -536.37kB  7.48%  k8s.io/apimachinery/pkg/api/meta.init
  518.02kB  7.22% 23.55%   518.02kB  7.22%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
     515kB  7.18% 16.38%      515kB  7.18%  regexp.makeOnePass.func1
     513kB  7.15%  9.23%      513kB  7.15%  golang.org/x/net/http2/hpack.newInternalNode
  512.38kB  7.14%  2.09%   512.38kB  7.14%  github.com/operator-framework/operator-registry/cmd/opm/index.newIndexDeprecateTruncateCmd
 -512.28kB  7.14%  9.23% -1626.82kB 22.67%  github.com/operator-framework/operator-registry/pkg/cache.(*JSON).Load
 -512.16kB  7.14% 16.36%  -512.16kB  7.14%  net.parseNSSConf
  512.05kB  7.14%  9.23%   512.05kB  7.14%  encoding/json.unquoteBytes
```

File: pkg/cache/json.go (hunk 2 above)
Comment (joelanford) [reply pkg/cache/json.go:87]: Sgtm. We can revisit later if it turns out not to be fine.

File: pkg/cache/json.go (hunk 1 above)
Comment (joelanford) [reply pkg/cache/json.go:105]: > If api.Bundle has an optional field and on index=0 we set it, but index=1 does not ... what does the content of the variable look like?

Right, we'd need to `.Reset()` it. But it looks like that just [re-initializes](https://github.com/operator-framework/operator-registry/blob/d02b0b68acab6bb218b431bed70b3703ffc168ab/pkg/api/registry.pb.go#L392) to an empty `Bundle`, so maybe unhelpful.

File: pkg/cache/json.go [hunk 3]
```diff
@@ -58,31 +60,66 @@ func (q *JSON) ListBundles(ctx context.Context) ([]*api.Bundle, error) {
...
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		if keys[i].chName != keys[j].chName {
+			return keys[i].chName < keys[j].chName
+		}
+		if keys[i].pkgName != keys[j].pkgName {
+			return keys[i].pkgName < keys[j].pkgName
+		}
+		return keys[i].name < keys[j].name
+	})
```
Comment (awgreene) [inline pkg/cache/json.go:79]: Why are we ordering the list here?

**Review by awgreene — COMMENTED** [review]
One question, looks good otherwise.

File: pkg/cache/json.go (hunk 3 above)
Comment (grokspawn) [reply pkg/cache/json.go:79]: For e.g. https://github.com/operator-framework/operator-registry/issues/1069

**Pull request merged**